File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No37_Sudoku_Solver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/28 20:10
# https://leetcode.com/problems/sudoku-solver/
from typing import List


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        self.solve(board)

    def solve(self, board):
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    for num in ["1", "2", "3", "4", "5", "6", "7", "8", "9"]:
                        if self.isValid(board, i, j, num):
                            board[i][j] = num
                            if self.solve(board):
                                return True
                            else:
                                board[i][j] = '.'
                    return False

        return True

    def isValid(self, board, row, col, num):
        for i in range(9):
            if board[row][i] == num or\
                    board[i][col] == num or \
                    board[(row // 3) * 3 + i // 3][(col // 3) * 3 + i % 3] == num:
                return False

        return True


if __name__ == '__main__':
    solution = Solution()

    board = [["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]
    expected = [["5","3","4","6","7","8","9","1","2"],["6","7","2","1","9","5","3","4","8"],["1","9","8","3","4","2","5","6","7"],["8","5","9","7","6","1","4","2","3"],["4","2","6","8","5","3","7","9","1"],["7","1","3","9","2","4","8","5","6"],["9","6","1","5","3","7","2","8","4"],["2","8","7","4","1","9","6","3","5"],["3","4","5","2","8","6","1","7","9"]]
    solution.solveSudoku(board)
    assert board == expected

File: src/main/java/com/hazza/leetcode/SegmentOne/No37SudokuSolver.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/sudoku-solver/#/description
 */
public class No37SudokuSolver {
    public void solveSudoku(char[][] board) {
        if (board == null || board.length ==0) return;
        solve(board);
    }

    private boolean solve(char[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    for (char c = '1'; c <= '9'; c++) {
                        if (isValid(board, i, j, c)) {
                            board[i][j] = c;
                            if (solve(board)) return true;
                            else board[i][j] = '.';     //go back
                        }
                    }
                    return false;
                }
            }
        }
        return true;

    }

    private boolean isValid(char board[][], int row, int col, char c) {
        for (int i = 0; i < 9; i++) {
            if (board[row][i] != '.' && board[row][i] == c) return false;
            if (board[i][col] != '.' && board[i][col] == c) return false;
            int rowIndex = (row / 3) * 3 + i / 3;
            int colIndex = (col / 3) * 3 + i % 3;
            if (board[rowIndex][colIndex] != '.' && board[rowIndex][colIndex] == c) return  false;
        }
        return true;
    }
}
